File: sdk/src/main/java/com/meniga/sdk/converters/TransactionEventSorter.java
package com.meniga.sdk.converters;

import com.meniga.sdk.models.feed.MenigaFeed;
import com.meniga.sdk.models.feed.MenigaFeedItem;
import com.meniga.sdk.models.feed.MenigaTransactionEvent;
import com.meniga.sdk.models.transactions.MenigaTransaction;

import java.util.ArrayList;
import java.util.List;

/**
 * Copyright 2018 Meniga Iceland Inc.
 * Created by agustk on 15.11.2018.
 */
public class TransactionEventSorter {

	TransactionEventSorter() {
	}

	void moveTransactionEventsToTransaction(MenigaFeed feed) {
		List<MenigaTransactionEvent> transactionEvents = new ArrayList<>();
		for (int i = 0; i < feed.size(); i++) {
			MenigaFeedItem item = feed.get(i);
			if (item instanceof MenigaTransactionEvent) {
				transactionEvents.add((MenigaTransactionEvent) item);
				feed.remove(item);
				i--;
			}
		}
		for (int i = 0; i < transactionEvents.size(); i++) {
			MenigaTransactionEvent item = transactionEvents.get(i);
			for (int x = 0; x < feed.size(); x++) {
				MenigaFeedItem feedItem = feed.get(x);
				if (feedItem instanceof MenigaTransaction && ((MenigaTransaction) feedItem).getId() == item.getTopicId()) {
					item.setDate(feedItem.getDate());
					feed.add(x + 1, item);
					x++;
				}
			}
		}
	}
}

File: sdk/src/main/java/com/meniga/sdk/converters/TransactionEventSorter.kt
package com.meniga.sdk.converters

import com.meniga.sdk.models.feed.MenigaFeed
import com.meniga.sdk.models.feed.MenigaTransactionEvent
import com.meniga.sdk.models.transactions.MenigaTransaction

/**
 * Copyright 2018 Meniga Iceland Inc.
 * Created by agustk on 16.11.2018.
 */
class TransactionEventSorter {

    fun moveTransactionEventsToTransaction(feed : MenigaFeed) {
        val transactionEvents = (0 until feed.size).map { feed[it] }
                .filterIsInstance(MenigaTransactionEvent::class.java)
                .reversed()

        feed.removeAll { it is MenigaTransactionEvent }

        transactionEvents.forEach { event ->
            val topicId = event.topicId;
            val subList = feed.filter { it is MenigaTransaction && it.id == topicId }

            if (subList.isEmpty()) {
                feed.find { it.date.millis > event.date.millis }.let {
                    if (it == null) {
                        feed.add(event)
                    } else {
                        feed.add(feed.indexOf(it), event)
                    }
                }
            } else {
                feed.add(feed.indexOf(subList.last()) + 1, event)
                event.setDate(subList.last().getDate())
            }
        }
    }
}
